csrc/ir/base_nodes.cpp

```diff
@@ -401,6 +401,21 @@ std::vector<PolymorphicValue> Expr::evaluate(
       "Please override the evaluate method");
 }
 
+std::vector<PolymorphicValue> Expr::evaluate(
+    const ExpressionEvaluator& ee,
+    std::unordered_map<const Val*, PolymorphicValue>& known_values) const {
+  std::vector<PolymorphicValue> expr_inputs;
+  expr_inputs.reserve(inputs().size());
+  for (auto inp : inputs()) {
+    const auto& eval_i = ee.evaluate(inp, known_values);
+    if (!eval_i.hasValue()) {
+      return {null_};
```

Comment:
Expr::null_ is probably unnecessary complexity. IIRC, all `std::monostate`s are equal. So you can simply say `return {std::monostate{}}` here. **If** the overhead of constructing these `monostate`s is indeed significant, you can store it in a static local variable, so it's only constructed once.